infra/Pos/Slotting/Util.hs

```diff
@@ -93,41 +97,85 @@ getNextEpochSlotDuration =
     esdSlotDuration . snd <$> getCurrentNextEpochSlottingDataM
 
 -- | Type constraint for `onNewSlot*` workers
-type OnNewSlot ctx m =
+type MonadOnNewSlot ctx m =
     ( MonadIO m
     , MonadReader ctx m
     , MonadSlots ctx m
     , MonadMask m
     , WithLogger m
+    , Mockable Async m
     , Mockable Delay m
     , MonadReporting ctx m
     , HasShutdownContext ctx
     , MonadRecoveryInfo m
     , HasConfiguration
     )
 
+-- | Parameters for `onNewSlot`.
+data OnNewSlotParams = OnNewSlotParams
+    { onspStartImmediately  :: !Bool
+    -- ^ Whether first action should be executed ASAP (i. e. basically
+    -- when the program starts), or only when new slot starts.
+    --
+    -- For example, if the program is started in the middle of a slot
+    -- and this parameter in 'False', we will wait for half of slot
+    -- and only then will do something.
+    , onspTerminationPolicy :: !ActionTerminationPolicy
+    -- ^ What should be done if given action doesn't finish before new
+    -- slot starts. See the description of 'ActionTerminationPolicy'.
+    }
+
+-- | Default parameters which were used by almost all code before this
+-- data type was introduced.
+instance Default OnNewSlotParams where
+    def =
+        OnNewSlotParams
+        { onspStartImmediately = True
+        , onspTerminationPolicy = NoTerminationPolicy
+        }
+
+-- | This policy specifies what should be done if the action passed to
+-- `onNewSlot` doesn't finish when current slot finishes.
+--
+-- We don't want to run given action more than once in parallel for
+-- variety of reasons:
+-- 1. If action hangs for some reason, there can be infinitely growing pool
+-- of hanging actions with probably bad consequences (e. g. leaking memory).
+-- 2. Thread management will be quite complicated if we want to fork
+-- threads inside `onNewSlot`.
+-- 3. If more than one action is launched, they may use same resources
+-- concurrently, so the code must account for it.
+data ActionTerminationPolicy
+    = NoTerminationPolicy
+    -- ^ Even if action keeps running after current slot finishes,
+    -- we'll just wait and start action again only after the previous
+    -- one finishes.
+    | NewSlotTerminationPolicy !Text
```

Comment:
Would passing a `LoggerName` be overkill or not what we want here?

Reply:
It's not what we want here, because this `Text` value is not used as a logger name, it's used as a part of the message which is logged (see below).
```
+                  logWarning $ sformat
+                  ("Action "%stext%
+                   " hasn't finished before new slot started") name
```